helpers/preview/bin/preview-rst

```diff
@@ -0,0 +1,313 @@
+        return 1
+    fi
+    rst_basename=`basename ${rst_file}`
+    # The output filename will be the same for every generation of the same
+    # input filename (to allow for browser refreshing)
+    get_output_dir=`get_output_dir '${rst_file}'`
+    html_file=`echo ${get_output_dir}/${rst_basename}.html | \
+        sed 's,\.rst,\.html,'`
+    # Build a standalone HTML file
+    pandoc_opts="--standalone --self-contained --fail-if-warnings"
+    pandoc ${pandoc_opts} -o "${html_file}" "${rst_file}"
+    if test ! -f "${html_file}"; then
+        echo "Could not generate output file: ${html_file}" >&2
+        return 1
+    fi
+    open "${html_file}"
+}
+
+# This function creates a temporary Sphinx project and runs `sphinx-autobuild`
+# to take advantage of live reloads (too much work to implement manually)
+function display_standalone_sphinx {
+    rst_file="${1}"
+    rst_basename=`basename "${rst_file}"`
+    # The output filename will be the same for every generation of the same
+    # input filename (to allow for browser refreshing)
+    sphinx_template_dir="${etc_dir}/sphinx"
+    output_dir=`get_output_dir "${rst_file}"`
+    sphinx_dir="${output_dir}/sphinx"
+    sphinx_rst_file="${sphinx_dir}/index.rst"
+    html_dir="${output_dir}/html"
+    log_dir="${output_dir}/log"
+    sphinx_log_file="${output_dir}/log/sphinx.log"
+    # We do not create `${sphinx_dir}` because this is done in a subsequent
+    # step (and doing it here would mess up the `cp` command)
+    mkdir -p "${html_dir}" "${log_dir}"
+    # Try to find any `sphinx-autobuild` running on port 8080
+    autobuild_pid=`ps aux | \
+        grep 'sphinx-autobuild --port 8080' | \
+        grep -v 'grep' | awk '{print $2}'`
+    if test ! -z "${autobuild_pid}"; then
+        # The `sphinx-autobuild` program is running, so we kill it before
+        # starting it up again.
+        #
+        # It doesn't matter if the existing `sphinx-autobuild` is running for
+        # this Sphinx directory or not because the user has requested a new
+        # preview, so it is reasonable to assume they want to start over. If
+        # they didn't want to start over, they can instead rely on the existing
+        # `sphinx-autobuild` to keep rebuilding the file.
+        kill "${autobuild_pid}"
+    fi
+    # Protective measure before copying (older versions of this directory may
+    # still exist)
+    rm -rf "${sphinx_dir}"
+    # Copy the template Sphinx project to the temporary directory
+    cp -R "${sphinx_template_dir}" "${sphinx_dir}"
+    # Soft link the provided RST file to the temporary Sphinx project
+    ln -s "${rst_file}" "${sphinx_rst_file}"
+    venv_dir=`get_venv_dir`
+    . "${venv_dir}/bin/activate"
+    # Run on port 8080 to avoid conflict with 8000 (the default port), which
+    # may be attached to an already running, but unrelated,`sphinx-autobuild`
+    nohup sphinx-autobuild \
+        --port 8080 --open-browser --delay 0 -b html \
```

Comment:
Port 8080 is a very common port which might already be used by other services, either on a typical development workstation or even on a production machine. We should choose a different one here. I suggest 24242.

Reply:
I wrote a quick function to get a random port in the dynamic range. there's no need for the port number to be consistent as the script opens the browser for you using the correct URL each time